commons-core/src/main/scala/com/avsystem/commons/rest/data.scala

```diff
@@ -0,0 +1,195 @@
+package com.avsystem.commons
+package rest
+
+import com.avsystem.commons.misc.{AbstractValueEnum, AbstractValueEnumCompanion, EnumCtx}
+import com.avsystem.commons.rpc._
+import com.avsystem.commons.serialization.GenCodec.ReadFailure
+import com.avsystem.commons.serialization.json.{JsonReader, JsonStringInput, JsonStringOutput}
+
+import scala.util.control.NoStackTrace
+
+sealed trait RestValue extends Any {
+  def value: String
+}
+
+/**
+  * Value used as encoding of [[Path]] parameters.
+  */
+case class PathValue(value: String) extends AnyVal with RestValue
+object PathValue {
+  def split(path: String): List[PathValue] =
+    path.split("/").iterator.filter(_.nonEmpty).map(PathValue(_)).toList
+}
+
+/**
+  * Value used as encoding of [[Header]] parameters.
+  */
+case class HeaderValue(value: String) extends AnyVal with RestValue
+
+/**
+  * Value used as encoding of [[Query]] parameters.
+  */
+case class QueryValue(value: String) extends AnyVal with RestValue
+
+/**
+  * Value used as encoding of [[JsonBodyParam]] parameters.
+  */
+case class JsonValue(value: String) extends AnyVal with RestValue
+
+/**
+  * Value used to represent HTTP body. Also used as direct encoding of [[Body]] parameters. Types that have
+  * encoding to [[JsonValue]] automatically have encoding to [[HttpBody]] which uses application/json MIME type.
+  * There is also a specialized encoding provided for `Unit` which returns empty HTTP body when writing and ignores
+  * the body when reading.
+  */
+sealed trait HttpBody {
+  def contentOpt: Opt[String] = this match {
+    case HttpBody(content, _) => Opt(content)
+    case HttpBody.Empty => Opt.Empty
+  }
+
+  def forNonEmpty(consumer: (String, String) => Unit): Unit = this match {
+    case HttpBody(content, mimeType) => consumer(content, mimeType)
+    case HttpBody.Empty =>
+  }
+
+  def readContent(): String = this match {
+    case HttpBody(content, _) => content
+    case HttpBody.Empty => throw new ReadFailure("Expected non-empty body")
+  }
+
+  def readJson(): JsonValue = this match {
+    case HttpBody(content, HttpBody.JsonType) => JsonValue(content)
+    case HttpBody(_, mimeType) =>
+      throw new ReadFailure(s"Expected body with application/json type, got $mimeType")
+    case HttpBody.Empty =>
+      throw new ReadFailure("Expected body with application/json type, got empty body")
+  }
+}
+object HttpBody {
+  object Empty extends HttpBody
+  final case class NonEmpty(content: String, mimeType: String) extends HttpBody
+
+  def empty: HttpBody = Empty
+
+  def apply(content: String, mimeType: String): HttpBody =
```

Comment:
Could you create a wrapper for `mimeType: String`?

Reply:
It seems a bit of overkill to me as it's already wrapped into `HttpBody` so it's meaning is clear and it's intended to be always used when handling bodies.